cmake/FlatccGenerateSources.cmake

```diff
@@ -0,0 +1,340 @@
+#
+#   EXTRA_ARGS <flatcc-arg> [<flatcc-arg> [...]]
+#
+#     Add extra arguments to flatcc.
+#     Use the arguments `ALL`/`COMMON`/`READER`/`JSON` instead of specifying `-a`/`-c`/`--reader`/`--json` here.
+#
+
+cmake_minimum_required(VERSION 3.3)
+
+# TODO: not needed when cmake_minimum_required_version >= 3.4
+include(CMakeParseArguments)
+
+function(flatcc_generate_sources)
+    # parse function arguments
+    set(output_options BINARY_SCHEMA COMMON COMMON_READER COMMON_BUILDER BUILDER READER VERIFIER JSON_PARSER JSON_PRINTER JSON)
+    set(NO_VAL_ARGS ALL RECURSIVE ${output_options})
+    set(SINGLE_VAL_ARGS NAME OUTPUT_DIR OUTFILE PREFIX)
+    set(MULTI_VAL_ARGS SCHEMA_FILES EXTRA_ARGS PATHS)
+
+    cmake_parse_arguments(FLATCC "${NO_VAL_ARGS}" "${SINGLE_VAL_ARGS}" "${MULTI_VAL_ARGS}" ${ARGN})
+    
+    # TODO: add following check when cmake_minimum_required_version >= 3.15
+    #if (FLATCC_KEYWORDS_MISSING_VALUES)
+    #    message(FATAL_ERROR "The following keywords in call to flatcc_generate_sources have no value: ${FLATCC_KEYWORDS_MISSING_VALUES}")
+    #endif()
+
+    if (FLATCC_UNPARSED_ARGUMENTS)
+        message(WARNING "The following unknown keywords in call to flatcc_generate_sources are ignored: ${FLATCC_UNPARSED_ARGUMENTS}")
+    endif()
+
+    if (NOT FLATCC_NAME)
+        message(FATAL_ERROR "NAME not provided in call to flatcc_generate_sources")
+    endif()
+    
+    if (NOT FLATCC_SCHEMA_FILES)
+        message(FATAL_ERROR "No flatbuffer schema files provided")
+    endif()
+
+    if (NOT FLATCC_OUTPUT_DIR)
+        set(FLATCC_OUTPUT_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+    endif()
+    if (NOT IS_ABSOLUTE "${FLATCC_OUTPUT_DIR}")
+        set(FLATCC_OUTPUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/${FLATCC_OUTPUT_DIR}")
+    endif()
+
+    list(APPEND FLATCC_ARGS -o "${FLATCC_OUTPUT_DIR}")
+
+    # Add current source directory for finding dependencies
+    set(absolute_flatcc_paths "${CMAKE_CURRENT_SOURCE_DIR}")
+
+    set(ABSOLUTE_SCHEMA_FILES)
+    foreach(schema_file ${FLATCC_SCHEMA_FILES})
+        # TODO: file(REAL_PATH) if cmake_minimum_required_version >= 3.19
+        if(IS_ABSOLUTE "${schema_file}")
+            set(absolute_def_file "${schema_file}")
+        else()
+            set(absolute_def_file "${CMAKE_CURRENT_SOURCE_DIR}/${schema_file}")
+        endif()
+        list(APPEND ABSOLUTE_SCHEMA_FILES "${absolute_def_file}")
+
+        get_filename_component(absolute_def_directory "${absolute_def_file}" DIRECTORY)
+        list(APPEND absolute_flatcc_paths "${absolute_def_directory}")
+    endforeach()
+
+    foreach(path ${FLATCC_PATHS})
+        # TODO: file(REAL_PATH) if cmake_minimum_required_version >= 3.19
+        if(NOT IS_ABSOLUTE "${path}")
+            set(path "${CMAKE_CURRENT_SOURCE_DIR}/${path}")
+        endif()
+        list(APPEND absolute_flatcc_paths "${path}")
+        list(APPEND FLATCC_ARGS -I "${path}")
+    endforeach()
```

Comment:
In this loop, all `PATHS` (=include paths) are converted to absolute paths.
The absolute paths are appended to an internal variable + added to the arguments (with `-I` prepended to each path).